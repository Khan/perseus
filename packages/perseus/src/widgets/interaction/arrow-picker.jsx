/* eslint-disable static-service/require-fixture */
// @flow

import * as React from "react";

import ButtonGroup from "../../../perseus-all-package/components/button-group.jsx";

import type {ChangeableProps} from "../../../perseus-all-package/mixins/changeable.jsx";

type Props = {|
    value?: string,
    ...ChangeableProps,
|};

class ArrowPicker extends React.Component<Props> {
    static defaultProps: {|value: string|} = {
        value: "",
    };

    render(): React.Node {
        return (
            <ButtonGroup
                value={this.props.value}
                allowEmpty={false}
                buttons={[
                    {value: "", content: <span>&mdash;</span>},
                    {value: "->", content: <span>&#x2192;</span>},
                    /*
                TODO(eater): fix khan-exercises so these are supported
                {value: "<-", content: <span>&#x2190;</span>},
                {value: "<->", content: <span>&#x2194;</span>}
                */
                ]}
                onChange={this.props.onChange}
            />
        );
    }
}

export default ArrowPicker;
