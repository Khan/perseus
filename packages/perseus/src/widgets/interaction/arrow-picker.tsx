import * as React from "react";

import ButtonGroup from "../../components/button-group";

import type {ChangeableProps} from "../../mixins/changeable";

type Props = {
    value?: string;
} & ChangeableProps;

class ArrowPicker extends React.Component<Props> {
    static defaultProps: {
        value: string;
    } = {
        value: "",
    };

    render(): React.ReactNode {
        return (
            <ButtonGroup
                value={this.props.value}
                allowEmpty={false}
                buttons={[
                    {value: "", content: <span>&mdash;</span>},
                    {value: "->", content: <span>&#x2192;</span>},
                    /*
                TODO(eater): fix khan-exercises so these are supported
                {value: "<-", content: <span>&#x2190;</span>},
                {value: "<->", content: <span>&#x2194;</span>}
                */
                ]}
                onChange={this.props.onChange}
            />
        );
    }
}

export default ArrowPicker;
