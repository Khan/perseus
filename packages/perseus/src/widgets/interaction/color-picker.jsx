/* eslint-disable static-service/require-fixture */
// @flow

import * as React from "react";
import _ from "underscore";

import ButtonGroup from "../../../perseus-all-package/components/button-group.jsx";
import KhanColors from "../../../perseus-all-package/util/colors.js";

import type {ChangeableProps} from "../../../perseus-all-package/mixins/changeable.jsx";

const COLORS = [
    KhanColors.BLACK,
    KhanColors.BLUE,
    KhanColors.GREEN,
    KhanColors.PINK,
    KhanColors.PURPLE,
    KhanColors.RED,
    KhanColors.GRAY,
];

const LIGHT_COLORS = [
    KhanColors.LIGHT_BLUE,
    KhanColors.LIGHT_ORANGE,
    KhanColors.LIGHT_PINK,
    KhanColors.LIGHT_GREEN,
    KhanColors.LIGHT_PURPLE,
    KhanColors.LIGHT_RED,
    "#fff",
];

type Props = {|
    lightColors?: boolean,
    value?: string,
    ...ChangeableProps,
|};

class ColorPicker extends React.Component<Props> {
    static defaultProps: {|lightColors: boolean, value: any|} = {
        value: KhanColors.BLACK,
        lightColors: false,
    };

    render(): React.Node {
        const colors = this.props.lightColors ? LIGHT_COLORS : COLORS;
        return (
            <ButtonGroup
                value={this.props.value}
                allowEmpty={false}
                buttons={_.map(colors, (color) => {
                    return {
                        value: color,
                        content: (
                            <span>
                                <span
                                    className="colorpicker-circle"
                                    style={{background: color}}
                                />
                                &nbsp;
                            </span>
                        ),
                    };
                })}
                onChange={this.props.onChange}
            />
        );
    }
}

export default ColorPicker;
