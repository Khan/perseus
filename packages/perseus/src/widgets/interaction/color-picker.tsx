import * as React from "react";
import _ from "underscore";

import ButtonGroup from "../../components/button-group";
import KhanColors from "../../util/colors";

import type {ChangeableProps} from "../../mixins/changeable";

const COLORS = [
    KhanColors.BLACK,
    KhanColors.BLUE,
    KhanColors.GREEN,
    KhanColors.PINK,
    KhanColors.PURPLE,
    KhanColors.RED,
    KhanColors.GRAY,
];

const LIGHT_COLORS = [
    KhanColors.LIGHT_BLUE,
    KhanColors.LIGHT_ORANGE,
    KhanColors.LIGHT_PINK,
    KhanColors.LIGHT_GREEN,
    KhanColors.LIGHT_PURPLE,
    KhanColors.LIGHT_RED,
    "#fff",
];

type Props = {
    lightColors?: boolean;
    value?: string;
} & ChangeableProps;

class ColorPicker extends React.Component<Props> {
    static defaultProps: {
        lightColors: boolean;
        value: any;
    } = {
        value: KhanColors.BLACK,
        lightColors: false,
    };

    render(): React.ReactNode {
        const colors = this.props.lightColors ? LIGHT_COLORS : COLORS;
        return (
            <ButtonGroup
                value={this.props.value}
                allowEmpty={false}
                buttons={_.map(colors, (color) => {
                    return {
                        value: color,
                        content: (
                            <span>
                                <span
                                    className="colorpicker-circle"
                                    style={{background: color}}
                                />
                                &nbsp;
                            </span>
                        ),
                    };
                })}
                onChange={this.props.onChange}
            />
        );
    }
}

export default ColorPicker;
