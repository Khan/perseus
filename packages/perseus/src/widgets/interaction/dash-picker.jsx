/* eslint-disable static-service/require-fixture */
// @flow

import * as React from "react";

import ButtonGroup from "../../../perseus-all-package/components/button-group.jsx";

import type {ChangeableProps} from "../../../perseus-all-package/mixins/changeable.jsx";

type Props = {|
    value?: string,
    ...ChangeableProps,
|};

class DashPicker extends React.Component<Props> {
    static defaultProps: {|value: string|} = {
        value: "",
    };

    render(): React.Node {
        return (
            <ButtonGroup
                value={this.props.value}
                allowEmpty={false}
                buttons={[
                    {value: "", content: <span>&mdash;</span>},
                    {value: "-", content: <span>&ndash;&ndash;&ndash;</span>},
                    {
                        value: "- ",
                        content: <span>&ndash;&nbsp;&nbsp;&ndash;</span>,
                    },
                    {
                        value: ".",
                        content: <span>&middot;&middot;&middot;&middot;</span>,
                    },
                    {
                        value: ". ",
                        content: <span>&middot; &middot; &middot;</span>,
                    },
                ]}
                onChange={this.props.onChange}
            />
        );
    }
}

export default DashPicker;
