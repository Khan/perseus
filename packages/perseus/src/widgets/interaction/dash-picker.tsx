import * as React from "react";

import ButtonGroup from "../../components/button-group";

import type {ChangeableProps} from "../../mixins/changeable";

type Props = {
    value?: string;
} & ChangeableProps;

class DashPicker extends React.Component<Props> {
    static defaultProps: {
        value: string;
    } = {
        value: "",
    };

    render(): React.ReactNode {
        return (
            <ButtonGroup
                value={this.props.value}
                allowEmpty={false}
                buttons={[
                    {value: "", content: <span>&mdash;</span>},
                    {value: "-", content: <span>&ndash;&ndash;&ndash;</span>},
                    {
                        value: "- ",
                        content: <span>&ndash;&nbsp;&nbsp;&ndash;</span>,
                    },
                    {
                        value: ".",
                        content: <span>&middot;&middot;&middot;&middot;</span>,
                    },
                    {
                        value: ". ",
                        content: <span>&middot; &middot; &middot;</span>,
                    },
                ]}
                onChange={this.props.onChange}
            />
        );
    }
}

export default DashPicker;
