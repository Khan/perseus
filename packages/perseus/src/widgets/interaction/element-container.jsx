/* eslint-disable jsx-a11y/anchor-is-valid, static-service/require-fixture */
// @flow
import * as React from "react";

import InlineIcon from "../../../perseus-all-package/components/inline-icon.jsx";
import {
    iconChevronDown,
    iconChevronRight,
    iconCircleArrowDown,
    iconCircleArrowUp,
    iconTrash,
} from "../../../perseus-all-package/icon-paths.js";

type ElementContainerProps = {|
    children: React.Element<any> | $ReadOnlyArray<React.Element<any>>,
    initiallyVisible: boolean,
    onDelete?: ?() => void,
    onDown?: ?() => void,
    onUp?: ?() => void,
    title: string | React.Element<any>,
|};

class ElementContainer extends React.Component<
    ElementContainerProps,
    {show: boolean, ...},
> {
    static defaultProps: {|initiallyVisible: boolean, title: string|} = {
        initiallyVisible: false,
        title: "More",
    };

    constructor(props: ElementContainerProps) {
        super(props);

        this.state = {
            show: props.initiallyVisible,
        };
    }

    toggle: (e: SyntheticEvent<>) => void = (e: SyntheticEvent<>) => {
        e.preventDefault();
        this.setState({show: !this.state.show});
    };

    render(): React.Element<"div"> {
        return (
            <div className="perseus-interaction-element">
                <a
                    href="#"
                    className={
                        "perseus-interaction-element-title " +
                        (this.state.show ? "open" : "closed")
                    }
                    onClick={this.toggle}
                >
                    {this.state.show ? (
                        <InlineIcon {...iconChevronDown} />
                    ) : (
                        <InlineIcon {...iconChevronRight} />
                    )}
                    {this.props.title}
                </a>
                <div
                    className={
                        "perseus-interaction-element-content " +
                        (this.state.show ? "enter" : "leave")
                    }
                >
                    {this.props.children}
                    {(this.props.onUp != null ||
                        this.props.onDown != null ||
                        this.props.onDelete != null) && (
                        <div className="edit-controls">
                            {this.props.onUp != null && (
                                <button onClick={this.props.onUp}>
                                    <InlineIcon {...iconCircleArrowUp} />
                                </button>
                            )}
                            {this.props.onDown != null && (
                                <button onClick={this.props.onDown}>
                                    <InlineIcon {...iconCircleArrowDown} />
                                </button>
                            )}
                            {this.props.onDelete != null && (
                                <button onClick={this.props.onDelete}>
                                    <InlineIcon {...iconTrash} />
                                </button>
                            )}
                        </div>
                    )}
                </div>
            </div>
        );
    }
}

export default ElementContainer;
