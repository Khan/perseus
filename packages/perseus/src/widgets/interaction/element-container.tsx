/* eslint-disable jsx-a11y/anchor-is-valid */
import * as React from "react";

import InlineIcon from "../../components/inline-icon";
import {
    iconChevronDown,
    iconChevronRight,
    iconCircleArrowDown,
    iconCircleArrowUp,
    iconTrash,
} from "../../icon-paths";

type Props = {
    children: React.ReactElement<any> | ReadonlyArray<React.ReactElement<any>>;
    initiallyVisible: boolean;
    onDelete?: () => void | null | undefined;
    onDown?: () => void | null | undefined;
    onUp?: () => void | null | undefined;
    title: string | React.ReactElement<any>;
};

type State = {
    show: boolean;
};

class ElementContainer extends React.Component<Props, State> {
    static defaultProps: {
        initiallyVisible: boolean;
        title: string;
    } = {
        initiallyVisible: false,
        title: "More",
    };

    constructor(props: Props) {
        super(props);

        this.state = {
            show: props.initiallyVisible,
        };
    }

    toggle: (e: React.SyntheticEvent) => void = (e: React.SyntheticEvent) => {
        e.preventDefault();
        this.setState({show: !this.state.show});
    };

    render(): React.ReactNode {
        return (
            <div className="perseus-interaction-element">
                <a
                    href="#"
                    className={
                        "perseus-interaction-element-title " +
                        (this.state.show ? "open" : "closed")
                    }
                    onClick={this.toggle}
                >
                    {this.state.show ? (
                        <InlineIcon {...iconChevronDown} />
                    ) : (
                        <InlineIcon {...iconChevronRight} />
                    )}
                    {this.props.title}
                </a>
                <div
                    className={
                        "perseus-interaction-element-content " +
                        (this.state.show ? "enter" : "leave")
                    }
                >
                    {this.props.children}
                    {(this.props.onUp != null ||
                        this.props.onDown != null ||
                        this.props.onDelete != null) && (
                        <div className="edit-controls">
                            {this.props.onUp != null && (
                                <button onClick={this.props.onUp}>
                                    <InlineIcon {...iconCircleArrowUp} />
                                </button>
                            )}
                            {this.props.onDown != null && (
                                <button onClick={this.props.onDown}>
                                    <InlineIcon {...iconCircleArrowDown} />
                                </button>
                            )}
                            {this.props.onDelete != null && (
                                <button onClick={this.props.onDelete}>
                                    <InlineIcon {...iconTrash} />
                                </button>
                            )}
                        </div>
                    )}
                </div>
            </div>
        );
    }
}

export default ElementContainer;
