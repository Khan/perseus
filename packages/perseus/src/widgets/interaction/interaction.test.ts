import {testDependencies} from "../../../../../testing/test-dependencies";
import {waitForInitialGraphieRender} from "../../../../../testing/wait";
import * as Dependencies from "../../dependencies";
import {renderQuestion} from "../__testutils__/renderQuestion";

import {question1} from "./interaction.testdata";

describe("interaction widget", () => {
    beforeEach(() => {
        jest.spyOn(Dependencies, "getDependencies").mockReturnValue(
            testDependencies,
        );
    });

    it("should render", async () => {
        // Arrange/Act
        const {container} = renderQuestion(question1);
        await waitForInitialGraphieRender();

        // Assert
        expect(container).toMatchSnapshot();
    });

    it("should be unanswerable", async () => {
        // Arrange/Act
        const {renderer} = renderQuestion(question1);
        await waitForInitialGraphieRender();

        // Assert
        // Note that this widget can never be answered correctly, no matter
        // what state its in.
        expect(renderer).toHaveBeenAnsweredIncorrectly();
    });
});
