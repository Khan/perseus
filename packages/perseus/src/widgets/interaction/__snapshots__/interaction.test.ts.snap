// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`interaction widget should render 1`] = `
<div>
  <div
    class="perseus-renderer perseus-renderer-responsive"
  >
    <div
      class="paragraph"
      data-perseus-paragraph-index="0"
    >
      <div
        class="paragraph"
      >
        Drag the dot all the way to the right.
      </div>
    </div>
    <div
      class="paragraph"
      data-perseus-paragraph-index="1"
    >
      <div
        class="paragraph"
      >
        <div
          class="perseus-widget-container widget-nohighlight widget-block"
        >
          <div
            class="graphie-container"
          >
            <div
              class="graphie"
              style="position: relative; width: 479.99999999999994px; height: 120.00000000000001px;"
            >
              <svg
                height="120.00000000000001"
                style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0); z-index: 2;"
                version="1.1"
                width="479.99999999999994"
                xmlns="http://www.w3.org/2000/svg"
              >
                <desc
                  style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0);"
                >
                  Created with Raphaël
                </desc>
                <defs
                  style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0);"
                />
              </svg>
              <svg
                height="120.00000000000001"
                style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0);"
                version="1.1"
                width="479.99999999999994"
                xmlns="http://www.w3.org/2000/svg"
              >
                <desc
                  style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0);"
                >
                  Created with Raphaël
                </desc>
                <defs
                  style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0);"
                />
                <path
                  d="M254.59459459459455,636.5217391304348L254.5785950953643,635.6596267341514L254.53065974038762,634.8009167016342L254.45097770861094,633.9489979690746L254.33986346861047,633.1072326705054L254.19775553752822,632.2789428689938L254.02521475044523,631.467397445977L253.82292204702173,630.6757992004796L253.5916757841394,629.9072722091297L253.33238858515142,629.1648494968558L253.04608373817518,628.4514610669225L252.73389115764147,627.7699223375465L252.39704292503845,627.1229230307273L252.0368684264488,626.513016557143L251.6547891060704,625.9426099390051L251.25231285642542,625.4139543106413L250.8310280673972,624.9291360342974L250.39259735758142,624.4900684662169L249.93875101268972,624.0984844055008L249.47128015690274,623.755929255539L248.99202968412115,623.4637549250093L248.5028909770123,623.2231144925131L248.0057944425869,623.0349576558989L247.5027018937646,622.9000269842405L246.99559880699437,622.8188549872548L246.48648648648643,622.7917620137299L245.9773741659785,622.8188549872548L245.4702710792083,622.9000269842405L244.96717853038598,623.0349576558989L244.47008199596056,623.2231144925131L243.9809432888517,623.4637549250093L243.50169281607012,623.755929255539L243.03422196028313,624.0984844055008L242.58037561539143,624.4900684662169L242.14194490557566,624.9291360342974L241.72066011654746,625.4139543106413L241.31818386690247,625.9426099390051L240.9361045465241,626.513016557143L240.57593004793443,627.1229230307273L240.23908181533136,627.7699223375465L239.92688923479767,628.4514610669225L239.64058438782146,629.1648494968558L239.38129718883349,629.9072722091297L239.15005092595115,630.6757992004796L238.94775822252765,631.467397445977L238.77521743544466,632.2789428689938L238.6331095043624,633.1072326705055L238.52199526436195,633.9489979690746L238.44231323258526,634.8009167016344L238.39437787760858,635.6596267341514L238.37837837837833,636.5217391304348L238.39437787760858,637.3838515267182L238.44231323258526,638.2425615592352L238.52199526436195,639.094480291795L238.6331095043624,639.9362455903641L238.77521743544466,640.7645353918757L238.94775822252765,641.5760808148926L239.15005092595115,642.3676790603901L239.3812971888335,643.1362060517399L239.64058438782146,643.8786287640137L239.9268892347977,644.592017193947L240.23908181533142,645.2735559233231L240.57593004793446,645.9205552301422L240.93610454652412,646.5304617037265L241.3181838669025,647.1008683218645L241.7206601165475,647.6295239502281L242.14194490557568,648.1143422265723L242.5803756153915,648.5534097946527L243.03422196028316,648.9449938553686L243.50169281607015,649.2875490053307L243.98094328885173,649.5797233358603L244.47008199596058,649.8203637683565L244.96717853038598,650.0085206049705L245.4702710792083,650.1434512766289L245.9773741659785,650.2246232736147L246.48648648648643,650.2517162471395L246.99559880699437,650.2246232736147L247.5027018937646,650.1434512766289L248.0057944425869,650.0085206049705L248.5028909770123,649.8203637683565L248.99202968412115,649.5797233358603L249.47128015690274,649.2875490053307L249.93875101268972,648.9449938553686L250.3925973575814,648.5534097946527L250.8310280673972,648.1143422265723L251.2523128564254,647.6295239502281L251.65478910607038,647.1008683218645L252.03686842644876,646.5304617037265L252.39704292503842,645.9205552301422L252.73389115764144,645.2735559233231L253.04608373817516,644.5920171939471L253.3323885851514,643.8786287640137L253.59167578413937,643.1362060517399L253.82292204702173,642.3676790603901L254.0252147504452,641.5760808148926L254.1977555375282,640.7645353918758L254.33986346861045,639.9362455903643L254.45097770861094,639.094480291795L254.53065974038762,638.2425615592355L254.5785950953643,637.3838515267184L254.59459459459455,636.5217391304349"
                  fill="none"
                  stroke="#6495ed"
                  stroke-dasharray="0"
                  stroke-width="4"
                  style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0); stroke-width: 4;"
                />
                <path
                  d="M19.459459459459445,62.60869565217392L19.459459459459445,114.78260869565217"
                  fill="none"
                  stroke="none"
                  stroke-dasharray="4,3"
                  stroke-width="1"
                  style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0); stroke-width: 1;"
                />
                <path
                  d="M84.3243243243243,62.60869565217392L84.3243243243243,114.78260869565217"
                  fill="none"
                  stroke="none"
                  stroke-dasharray="4,3"
                  stroke-width="1"
                  style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0); stroke-width: 1;"
                />
                <path
                  d="M149.18918918918916,62.60869565217392L149.18918918918916,114.78260869565217"
                  fill="none"
                  stroke="none"
                  stroke-dasharray="4,3"
                  stroke-width="1"
                  style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0); stroke-width: 1;"
                />
                <path
                  d="M214.054054054054,62.60869565217392L214.054054054054,114.78260869565217"
                  fill="none"
                  stroke="none"
                  stroke-dasharray="4,3"
                  stroke-width="1"
                  style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0); stroke-width: 1;"
                />
                <path
                  d="M278.91891891891885,62.60869565217392L278.91891891891885,114.78260869565217"
                  fill="none"
                  stroke="none"
                  stroke-dasharray="4,3"
                  stroke-width="1"
                  style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0); stroke-width: 1;"
                />
                <path
                  d="M343.78378378378375,62.60869565217392L343.78378378378375,114.78260869565217"
                  fill="none"
                  stroke="none"
                  stroke-dasharray="4,3"
                  stroke-width="1"
                  style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0); stroke-width: 1;"
                />
                <path
                  d="M408.6486486486486,62.60869565217392L408.6486486486486,114.78260869565217"
                  fill="none"
                  stroke="none"
                  stroke-dasharray="4,3"
                  stroke-width="1"
                  style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0); stroke-width: 1;"
                />
                <path
                  d="M473.51351351351343,62.60869565217392L473.51351351351343,114.78260869565217"
                  fill="none"
                  stroke="none"
                  stroke-dasharray="4,3"
                  stroke-width="1"
                  style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0); stroke-width: 1;"
                />
                <path
                  d="M19.459459459459445,10.434782608695661L149.18918918918916,10.434782608695661"
                  fill="none"
                  stroke="none"
                  stroke-dasharray="0"
                  stroke-width="2"
                  style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0); stroke-width: 2;"
                />
                <ellipse
                  cx="278.91891891891885"
                  cy="104.34782608695652"
                  fill="none"
                  rx="4"
                  ry="4"
                  stroke="none"
                  stroke-width="2"
                  style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0); stroke-width: 2;"
                />
              </svg>
              <div
                style="position: absolute; left: 0px; top: 0px;"
              >
                <div
                  style="position: absolute; width: 26px; height: 26px; left: 6.459459459459445px; top: -2.565217391304339px; transform: translateX(0px) translateY(0px) translateZ(0);"
                >
                  <svg
                    height="26"
                    style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0);"
                    version="1.1"
                    width="26"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <desc
                      style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0);"
                    >
                      Created with Raphaël
                    </desc>
                    <defs
                      style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0);"
                    />
                    <ellipse
                      cx="NaN"
                      cy="NaN"
                      fill="#71b307"
                      r="NaN"
                      rx="4"
                      ry="4"
                      stroke="#71b307"
                      style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0);"
                      transform=""
                    />
                  </svg>
                </div>
              </div>
              <div
                style="position: absolute; left: 0px; top: 0px; z-index: 2;"
              >
                <div
                  data-interactive-kind-for-testing="movable-point"
                  style="position: absolute; width: 48px; height: 48px; left: -4.540540540540555px; top: -13.565217391304339px; transform: translateX(0px) translateY(0px) translateZ(0);"
                >
                  <svg
                    height="48"
                    style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0);"
                    version="1.1"
                    width="48"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <desc
                      style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0);"
                    >
                      Created with Raphaël
                    </desc>
                    <defs
                      style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0);"
                    />
                    <ellipse
                      cx="24"
                      cy="24"
                      fill="#000000"
                      opacity="0"
                      rx="24"
                      ry="24"
                      stroke="#000"
                      style="-webkit-tap-highlight-color: rgba(0, 0, 0, 0); opacity: 0; cursor: move;"
                    />
                  </svg>
                </div>
              </div>
              <span
                class="graphie-label"
                data-math-formula="\\Huge 2"
                style="position: absolute; padding: 7px; color: rgb(100, 149, 237); left: 116.59637038983988px; top: 88.69565217391305px; fill: none;"
              >
                <span
                  class="tex-holder"
                />
              </span>
              <span
                class="graphie-label"
                data-math-formula="\\Huge 5"
                style="position: absolute; padding: 7px; color: rgb(100, 149, 237); left: 181.46123525470475px; top: 88.69565217391305px; fill: none;"
              >
                <span
                  class="tex-holder"
                />
              </span>
              <span
                class="graphie-label"
                data-math-formula="\\Huge 9"
                style="position: absolute; padding: 7px; color: rgb(100, 149, 237); left: 246.3261001195696px; top: 88.69565217391305px; fill: none;"
              >
                <span
                  class="tex-holder"
                />
              </span>
              <span
                class="graphie-label"
                data-math-formula="\\small 259"
                style="position: absolute; padding: 7px; color: gray; left: 19.459459459459445px; top: 26.08695652173914px; fill: none;"
              >
                <span
                  class="tex-holder"
                />
              </span>
              <span
                class="graphie-label"
                data-math-formula="\\small 259 \\times 10"
                style="position: absolute; padding: 7px; color: gray; left: 149.18918918918916px; top: 26.08695652173914px; fill: none;"
              >
                <span
                  class="tex-holder"
                />
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="paragraph"
      data-perseus-paragraph-index="2"
    >
      <div
        class="paragraph"
      >
        <em>
          Notice that we add a zero to the empty place value.
        </em>
         
      </div>
    </div>
  </div>
</div>
`;
